### ISSUE TITLE
Web annotator for content - MVP

### ISSUE THREAD

**Issue by ejhumphrey** [cosmir/open-mic#8]
Blocked on #10 (submission endpoint)
Blocked on @justinsalamon & possibly open-sourcing webtools?

Deliverable
- [x] Plays audio in the browser (hard coded)
- [x] Populates a number of instrument tags (hard coded)
- [x] Posts submission to backend

**Comment by dmcennis** [comment]
For clarity - what am I missing from this list of expanded requirements

- [ ] Audio data distributed to the user in a Replayable / seekable widget with visible timer
- [ ] Visual representation of the audio data itself (FFT?)
- [ ] Capacity to mark the visual representation
- [ ] Sync line in the visual display matching the audio timer
- [ ] Show time mark on hover on all marks on the audio visualization
- [ ] Plugin location for javascript powered widgets for defining the meaning of marks and/or additional annotation display/edits specific to a given task. Could include disabling markers ie. genre classification.
- [ ] interface for evaluation task specific widgets

**Comment by ejhumphrey** [comment]
awesome! two quick thoughts to add to the thread:

1. There are a lot of good ideas in that list. That said, the user interface requirements of the annotation tool will be dictated by the problem definition, raised by @julian-urbano in #2 (IIRC), so that we don't build more than we need (playback, visualization, yes; seekable playback, marking the representation, task-specific tools, probably not). On top of that, we'll hopefully be able to piggyback the work of @justinsalamon and SONYC, since they've already built a pretty shiny web annotator that already does more than we need. 
![image](https://cloud.githubusercontent.com/assets/926379/18131449/e6d8089e-6f60-11e6-98f4-a2d54eddd63f.png)

2. I'm going to be a stickler about "Build-Improve" -- we should challenge ourselves to build the smallest pieces of functionality that get us to the next step. We're a big group, and it's our only hope for parallelizing the work in a sane and productive way, so we can make sure all the pieces keep playing nicely together... otherwise we're going to end up like the [SNCF](http://www.bbc.com/news/world-europe-27497727).

**Comment by jtcramer** [comment]
Our team at Gracenote also has an annotation tool (built for the task of structural segmentation) that we are contemplating making open source.

Here's a screenshot:

<img width="945" alt="screen shot 2016-09-05 at 11 58 31 pm" src="https://cloud.githubusercontent.com/assets/1388668/18264479/474cd94e-73c5-11e6-9884-6741e4756205.png">

If it would help, we could also leverage parts of this tool in building an annotation tool for this project.

**Comment by stefan-balke** [comment]
@jtcramer awesome. What JS-Tools did you use and how did you do the audio playback (HTML5 or WebAudio)? Are the features pre-computed or everything in JavaScript?

**Comment by jtcramer** [comment]
The tool is primarily built with [wavesurfer.js](https://wavesurfer-js.org), jQuery, and Bootstrap. wavesurfer.js uses both HTML5 or WebAudio (which can be configured). The spectrogram is computed in Javascript using a plugin. wavesurfer.js is designed to be extended with plugins, which makes it really nice and easy to use.

**Comment by constd** [comment]
While I'd root for wavesurfer.js (because it's easy to use and expand) and hopefully the annotator @jtcramer showed, I'm throwing another option that's coming from BBC and is called [Peaks.js](https://github.com/bbcrd/peaks.js). 
![peaksjs](https://cloud.githubusercontent.com/assets/6237610/18348593/843d98d0-75cc-11e6-8407-8594f8eb5166.png)

It certainly looks interesting and was made for annotations, which is what we're looking for. It does not display spectrograms, though.

**Comment by justinsalamon** [comment]
Our annotator is also based on wavesurfer. Team is on board for open sourcing it, just need to separate the front-end code from the back-end code (currently slightly entangled). Should have en ETA soon.

**Comment by justinsalamon** [comment]
Our annotator is also based on wavesurfer. Team is on board for open sourcing it, just need to separate the front-end code from the back-end code (currently slightly entangled). Should have en ETA soon.

**Comment by bmcfee** [comment]
:+1: for wavesurfer.  I've worked a little with peaks, and it's nice, but pretty unwieldy if you want to modify the controls.

@jtcramer that interface looks fantastic, and almost exactly what I had in mind.  Thinking long-term, I'd like to support both strict categorical annotations (like you have shown) and multi-label annotations; this should be relatively easy.  I'd also like to have a feedback mechanism for annotators to flag an example as "bad" somehow -- non-musical (eg crowd noise), out-of-gamut (eg, chord annotations for a spoken word recording), or corrupt example.  I think these should be relatively easy to tack on, and should not hold up progress on the first round.

One question though: do we really want to show spectrograms to annotators?  I've heard concerns voiced in the past that this additional stimulus can bias annotators toward certain types of behaviors, so maybe we should be careful about that?  We can maybe pick that up in a side issue.

**Comment by ejhumphrey** [comment]
> [re:viz] We can maybe pick that up in a side issue.

👍  yep! 

There's a lot of good discussion of tools / prior work in this thread, and I'm hoping to steer it back to two goals: one, identify the checkboxes for an MVP annotator, which I think are...
- audio playback -- play button only, no visualization, playback cursor / control. could hard-code an audio file (on the web) for v0
- select / deselect a single tag, e.g. guitar
- submit button POSTs JSON to the `annotation/submit` endpoint

(any minimal functionality missing here?)

two, what's the fastest way to get there: SONYC's annotator, Gracenote's annotator, wavesurfer / peaks + effort, etc.
@justinsalamon mentioned working on an ETA for open-sourcing the SONYC annotator; @jtcramer, any idea what / how long it'd take to do the same with the one you mentioned?

there are also some issues I see spawning from this awesome thread:
- to visualize or not to visualize
- tag taxonomy presentation layout, e.g. grouped hierarchically, string matching text field a la search box, something else entirely?
- annotator feedback wishlist, e.g. out-of-gamut track for a task, different input paradigms, task switching

**Comment by jtcramer** [comment]
While peaks.js looks like it could be promising, I'd argue it may be better to go with wavesurfer since many of us already have experience with wavesurfer.js (I believe @constd has also used it). Plus we already have two annotation tools we could use and can potentially mash up parts between them if we need to.

> > [re:viz] We can maybe pick that up in a side issue.
> 👍 yep!
👍 agreed, made an issue here: #20 

@bmcfee I'm all for annotators flagging examples. Shouldn't be difficult to add that in.

@ejhumphrey The MVP checkpoints sound reasonable to me. Still trying to figure out how long it will take to open source our tool, but I'll get back to y'all as soon as I get an update.

We were planning on having a feedback cycle (regarding tool usability) with annotators and were planning on asking whether having something like a spectrogram would actually be helpful or not. We could certainly get incremental feedback from small user studies if that's feasible.

On a related note, the tool also integrates with Slack to facilitate the process of notifying users, getting feedback from users, and having an open communication channel between the annotators and the ones assigning tasks. Would something like this be valuable to the project?

**Comment by julian-urbano** [comment]
:thumbsup: to the feedback. Always. Two types are usually useful: i) comments about this example (doesn't work, etc) and ii) comments about the task in general (i.e. how to improve). It's always surprising to see what people say when you let them.

Left comments on visualization in #20

**Comment by ejhumphrey** [comment]
hey annotation team, couple things:

- any updates on open sourcing the annotation tools? @jtcramer @justinsalamon? I'm specifically interested in timelines -- I'd like to avoid reinventing the wheel if we can, but at a certain point, we will have to start making progress.
- it's recently come to my attention that github pages is wicked easy and that we can run our entire user-facing web stuff out of this repository. Check it: https://cosmir.github.io/open-mic/ Does anyone have any experience with this? Is it as awesome as it seems? @stefan-balke, thoughts on how this might work for #11?

**Comment by justinsalamon** [comment]
> any updates on open sourcing the annotation tools? @jtcramer @justinsalamon? I'm specifically interested in timelines -- I'd like to avoid reinventing the wheel if we can, but at a certain point, we will have to start making progress.

I can get you an ETA on Tuesday hopefully.

**Comment by justinsalamon** [comment]
Alright, so that didn't happen by Tuesday, but our interface is live and open-source:

https://github.com/CrowdCurio/audio-annotator

Fork time?

**Comment by ejhumphrey** [comment]
check it out @justinsalamon 
![screenshot 2016-11-19 12 55 01](https://cloud.githubusercontent.com/assets/926379/20457356/83cd2b40-ae57-11e6-9811-a2240193da0b.png)

pretty much got it hacked out ... currently on my branch, I've got the audio-annotator machinery talking to the flask server:
- annotator asks the server for a task
- server randomly identifies a (previously ingested) URI, and provides the appropriate URL to GET
- server fetches the taxonomy dynamically from github
- annotator does what it should with the task data
- submit gets thrown against a real backend endpoint
- rinse; repeat

couple more tweaks to clean up, but this thing is basically working end-to-end. Also this web front-end is goddamn sweet. +1000000 karma points

**Comment by justinsalamon** [comment]
@ejhumphrey awesomeness!

**Comment by ffont** [comment]
This looks really nice!
Where does this get the audio from? 
Is the spectrogram computed in the browser or does the backend server need to do that?
With a quick look at the code it seems this is client-side which is cool, but just to confirm...

Maybe we could try ingesting some of the Jamendo collection to the system?

**Comment by ejhumphrey** [comment]
The audio is coming from the flask server, running locally at the moment. The frontend talks to a [task endpoint](https://github.com/cosmir/open-mic/blob/ejh_20161119_iss8_webannot/audio-annotator/index.html#L72), which returns JSON that looks like ... 
```
{"task": {"url": "http://localhost:8080/audio/{uri}", ...}}
```
with the audio URL itself being another endpoint on the server. This allows us to build logic into the server to decide what content needs annotating, based on the user, overall progress, etc.

Spectrogram is computed in the browser. @justinsalamon mentioned that there's a way to visualize the waveform, but I couldn't figure it out quickly. I was also thinking we could distribute spectrogram computation and send back data on submission for more quickly rendering images in the future, but it's pretty fast as is and might not be necessary.

So ya, I think the next steps are (1) finish the [PR on the backend server](https://github.com/cosmir/open-mic/pull/25), (2) get this current thing buttoned up into a proper PR and tested, and (3) build-up and test integration with GCP / App Engine.

**Comment by ejhumphrey** [comment]
this is closed as of c3f72bcc7e503ba0e867f579d23600a0f0f36ada
